### ISSUE TITLE
Addon-docs: support interactions with existing addons

### ISSUE THREAD

**Issue by shilman** [storybooks/storybook#6700]

**Comment by rwieruch** [comment]
Just set up Docs today. Wonderful work that you have done there!

I would love if the old "Canvas" Tab would disappear and the new "Docs" Tab would be the one and only. Then the Tabbar could be removed and would declutter the UI. However, then it would be possible to integrate all the other addons (Knobs etc.) in the "Docs" Tab.

Since it wouldn't be possible anymore to open a single story via the left hand-side navigation, browsing between stories of one component should scroll you up and down in the "Docs" Tab to the related story.

Hope my input helps here. Thanks again for all your hard work!

**Comment by shilman** [comment]
Thanks @rwieruch! 🙏 

You can get part of the behavior you're looking for by running `start-storybook` or `build-storybook` with the `--docs` flag. This is still WIP, and will be getting some improvements in the next few days, but should give you a picture of what's to come.

Supporting all the addons is a longer-term project though. So until we figure that out, it's an either/or situation. 😺

**Comment by leoyli** [comment]
Agreed.  From what I've seen, there are isn't good agreement on how to structural an addon at first place, and potentially multi-framework support will makes the situation more challenging...

**Comment by techieshark** [comment]
I wonder how people handle knobs + docs addons. 

Specifically I'm wondering what folks do about the fact that all the "knobbed" component properties show up in the "show code" part of addon-docs. Presumably, doc-readers aren't interested in the knob functions (`boolean('disabled', true)` etc).

**Comment by shilman** [comment]
@techieshark There's a wonderful proof of concept from @JeroenReumkens that adds knobs to the prop table in docs mode. I'm not sure how it would relate to Storybook's existing knobs setup. :(

**Comment by techieshark** [comment]
@shilman that would be amazing - that's pretty much how I use knobs right now (one story for all the knobs, then specific stories to doc particular use cases). @JeroenReumkens I'd be keen to check that out.

**Comment by mmartinsky** [comment]
This would be really helpful for using addons like https://github.com/ticketmaster/storybook-styled-components for swapping library S-C/Emotion themes (not global Storybook theme).

**Comment by fabiradi** [comment]
Canvas and Docs both deserve their space and both have their own use cases. I wouldn't generally try to mangle them together.

**Docs:** Before Docs, I used the first Story to give an overview, explain a little bit about the component. The following stories give examples and show specific situations. The last story has interactions (knobs) to tweak the component in various ways. Docs now combines these views in a very nice and clean way (but without any addons).

**Canvas:** There are plenty of addons that work in canvas mode on a deeper, sometimes technical, level. I wouldn't want them to appear in Docs mode. 

### :thinking: Just an idea ...

I still find it a little confusing, to switch between Docs/Canvas in the tab bar. It is not very obvious to new readers. I had to explain it to a good number of people. On top, selecting a story in the navigation does not always have an effect (story already scrolled into view, etc.)

 Maybe the navigation could work this way:

- The **story kind** (CSF default export) opens Docs view. Currently, it opens the first story – which is still better than before, when it did not open anything.
- **Each story** opens the story in Canvas mode. This seems to be a more obvious target, if the user clicks on a specific story.

This would enable:

- :tada: **Contexts in Docs mode** applied to each story panel (themes, i18n, backgrounds ...). Addon-context is a great add-on that works much better than many panel-based addons.
- :framed_picture: **Viewports** for each story panel could actually work quite well.
- :dash: **Getting rid of the switcher** Canvas/Docs in the tab bar would create cleaner, consistent look.

I'm really missing **Contexts in Docs mode**. At least with React, it should be possible to set a  context and have it applied to each story panel (Themes, i18n, ...)

:man_shrugging: I'm not sure if any of this makes sense. These are just my experiences and thoughts.

**Comment by shilman** [comment]
@fabiradi great feedback. still chewing on this -- i think it's going to take some time to resolve

**Comment by sanchez** [comment]
Knobs and Actions could be added on a per story basis, so on the docs page you have a "Controls" and "Actions" button the bottom left (similar to the code button on the bottom right). Triggering this buttons unfolds a small section where the user can use Knobs and Actions as they would in the past (similar again to how the code snippet folds out)

**Comment by shilman** [comment]
Lol @sanchez 😂

**Comment by khantopa** [comment]
> I wonder how people handle knobs + docs addons.
> 
> Specifically I'm wondering what folks do about the fact that all the "knobbed" component properties show up in the "show code" part of addon-docs. Presumably, doc-readers aren't interested in the knob functions (`boolean('disabled', true)` etc).

Looking for the same question

**Comment by bolora** [comment]
> > I wonder how people handle knobs + docs addons.
> > Specifically I'm wondering what folks do about the fact that all the "knobbed" component properties show up in the "show code" part of addon-docs. Presumably, doc-readers aren't interested in the knob functions (`boolean('disabled', true)` etc).
> 
> Looking for the same question

I have yet to see a solution for this. It honestly makes copy code USELESS. Can someone please answer this question?

**Comment by josephfusco** [comment]
@bolora You can explicitly set the code that goes in the **show code** panel by using `Preview`'s `mdxSource` prop.

I'm not sure what the code would look like for the CSF or storiesOf formats however this is what it would look like for the [newer MDX stories format](https://github.com/storybookjs/storybook/blob/next/addons/docs/docs/mdx.md):
```js
<Preview mdxSource={`
  <div class="form-group">
    <label for="input-2" class="sr-only">...</label>
    <input id="input-2" type="text" class="form-control" placeholder="...">
  </div>
`}>
  <Story name="Text Input" height="100px">
    {`
      <div class="form-group">
        <label for="input-1" class="sr-only">Label text</label>
        <input id="input-1" type="text" class="form-control" placeholder="placeholder text">
      </div>
    `}
  </Story>
</Preview>
```

**Comment by jack-bliss** [comment]
Sincerely hope this is still being worked on - the --docs option is kinda useless without support for other addons! I think for the knobs/actions case, what @sanchez suggested is perfect, just make them per-story! Also the same can be said for viewports, just let me declare a default viewport using the normal parameters prop and have the inline story render in that viewport. This would make --docs my default option! For now i have to add "click the canvas tab above" for all my mobile stories.

**Comment by shilman** [comment]
@jack-bliss if it's useless to you, don't use it

**Comment by shilman** [comment]
@fabiradi finally getting back to this issue. addon-contexts is being replaced by addon-toolbars, which is docs-compatible. please take a look and let me know what you think: https://github.com/storybookjs/storybook/tree/next/addons/toolbars

**Comment by gaurav5430** [comment]
I can see that this is being worked on, which is great :)

Just wanted to confirm if it is possible yet to show a story inside a particular viewport when in the Docs view ?

for example, I have set the default viewports for some of my stories, but that disappears when i switch to the docs page. In this particular case, i don't really need the viewport button in the docs page, i just need a way/workaround to show the story in a particular viewport when rendered inside the docs page

**Comment by shilman** [comment]
@gaurav5430 not currently. we've been making progress though and I think we can probably get it into 6.1

**Comment by gaurav5430** [comment]
@shilman that would be great :)

Meanwhile, do you think there is a convenient workaround to achieve this?

just thinking out loud, i can possibly render the story inside an iframe and set the size of the iframe to whatever viewport i want, so that it displays as expected in the docs page (but then i would have to forego using the actual viewports addon and default viewport on the canvas view as well)

**Comment by floustao** [comment]
> @bolora You can explicitly set the code that goes in the **show code** panel by using `Preview`'s `mdxSource` prop.
> 
> I'm not sure what the code would look like for the CSF or storiesOf formats however this is what it would look like for the [newer MDX stories format](https://github.com/storybookjs/storybook/blob/next/addons/docs/docs/mdx.md):
> 
> ```js
> <Preview mdxSource={`
>   <div class="form-group">
>     <label for="input-1" class="sr-only">...</label>
>     <input id="input-1" type="text" class="form-control" placeholder="...">
>   </div>
> `}>
>   <Story name="Text Input" height="100px">
>     {`
>       <div class="form-group">
>         <label for="input-1" class="sr-only">Label text</label>
>         <input id="input-1" type="text" class="form-control" placeholder="placeholder text">
>       </div>
>     `}
>   </Story>
> </Preview>
> ```

This is awesome, thanks for sharing @josephfusco, I could not find it in the documentation.

**Comment by electriccode** [comment]
> @shilman that would be great :)
> 
> Meanwhile, do you think there is a convenient workaround to achieve this?
> 
> just thinking out loud, i can possibly render the story inside an iframe and set the size of the iframe to whatever viewport i want, so that it displays as expected in the docs page (but then i would have to forego using the actual viewports addon and default viewport on the canvas view as well)

I also have the same use case. Right now `<Story />` from docs addon doesn't accept `width` attribute so this is not possible.

**Comment by florian-milky** [comment]
If you wrap it into `<Preview>` or `<Canvas>` you can apply `width` style there:
```
<Preview style={INITIAL_VIEWPORTS.iphone6.styles}>
  <Story
    name="iphone6"
    parameters={{
      docs: {
        inlineStories: false,
        iframeHeight: INITIAL_VIEWPORTS.iphone6.styles.height,
      },
      layout: 'fullscreen',
      viewport: {
        viewports: INITIAL_VIEWPORTS,
        defaultViewport: 'iphone6',
      },
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>
```

**Comment by kremerd** [comment]
Is this topic still open for discussion?

I maintain an addon that does not depend on the currently visible stori(es) at all, but only interacts with global state (to switch themes). Other addons like the official "Actions" addon might be easily extendable to support a list of stories rather than a single story. It would be cool to add support for such usecases eventually.
